[TensorExpr] Properly handle input types promotion and special case of empty inputs for aten::cat.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #46447 [TensorExpr] Re-enable test for torch.cat, add a test for torch.cat being a single node in a fusion group.
* **#46500 [TensorExpr] Properly handle input types promotion and special case of empty inputs for aten::cat.**
* #46482 [TensorExpr] Fix shape inference logic for aten::cat.

Differential Revision: [D24373671](https://our.internmc.facebook.com/intern/diff/D24373671)

Commit:
[TensorExpr] Properly handle input types promotion and special case of empty inputs for aten::cat.

[ghstack-poisoned]

ngimel commented on test/cpp/tensorexpr/test_kernel.cpp:
```diff
@@ -353,6 +353,92 @@ void testKernel_4() {
+    auto a = at::rand({5, 3, 2}, TensorOptions(kCPU).dtype(at::kFloat));
+    auto b = at::rand({5, 7, 2}, TensorOptions(kCPU).dtype(at::kHalf));
+    auto c = at::rand({5, 9, 2}, TensorOptions(kCPU).dtype(at::kDouble));
+    auto o = at::zeros({5, 19, 2}, TensorOptions(kCPU).dtype(at::kDouble));
```
I don't see where this `o` is used - it is later assigned to the output of compiled function, but this initialization seems to be in vain?

ZolotukhinM replied:
Good catch, we don't need it!

Commit:
Update on "[TensorExpr] Properly handle input types promotion and special case of empty inputs for aten::cat."


Differential Revision: [D24373671](https://our.internmc.facebook.com/intern/diff/D24373671)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Properly handle input types promotion and special case of empty inputs for aten::cat."


Differential Revision: [D24373671](https://our.internmc.facebook.com/intern/diff/D24373671)

[ghstack-poisoned]
